File: web-ui/src/services/api/review.ts
import { api } from '../api';

/**
 * `/api/v1/review` — everything finPal guessed, in one place.
 *
 * *** THE COUNTS COME FROM THE SERVER AND ARE NEVER DERIVED HERE. *** A confirm
 * answers with the whole refreshed page, so the badge and the list are always
 * the same number from the same query. Decrementing a local count after a
 * successful POST is the shape where two open tabs disagree about how much is
 * left to do — D-101, one layer up.
 *
 * *** THERE IS NO `limit` ON THE WIRE, AND NOTHING HERE MAY ADD ONE. *** The
 * server sends the honest count and a capped list on purpose: "3 to review",
 * never "3 of 47". A page size would let a client reconstruct the denominator
 * this page exists without.
 *
 * *** NO TRAILING SLASH. *** `/api/v1/review`, matching the rest of web-ui — a
 * trailing slash has already split this API once, with web-ui omitting it and
 * mobile including it.
 */

/** A category whose spending group finPal guessed. */
export interface ReviewCategoryRow {
  id: number;
  name: string;
  parent_name: string | null;
  spending_type: string | null;
  /** Why finPal guessed. A guess rendered without its reason asks the user to
   *  re-derive the question (D-77, D-108). */
  reason: string;
}

/** An account whose type finPal inferred rather than being told (D-191). */
export interface ReviewAccountRow {
  id: number;
  name: string;
  type: string | null;
  balance: number | null;
  reason: string;
}

/** A transaction with no category. finPal has no opinion about these. */
export interface ReviewTransactionRow {
  id: number;
  description: string | null;
  amount: number | null;
  currency_code: string | null;
  date: string | null;
  transaction_type: string | null;
}

/**
 * The verb a section takes.
 *
 * *** `choose` IS NOT `confirm` AND MUST NOT BE RENDERED AS ONE. *** The first
 * two sections ask "finPal decided X — was it right?"; the third says finPal has
 * no opinion, so there is nothing to agree with. Giving an uncategorised
 * transaction a Confirm button would be a control that cannot mean anything.
 */
export type ReviewAction = 'confirm' | 'choose';

export interface ReviewSection<Row> {
  action: ReviewAction;
  rows: Row[];
}

export interface ReviewPayload {
  total: number;
  counts: { categories: number; accounts: number; uncategorised: number };
  sections: {
    categories: ReviewSection<ReviewCategoryRow>;
    accounts: ReviewSection<ReviewAccountRow>;
    uncategorised: ReviewSection<ReviewTransactionRow>;
  };
  /** Present only on a confirm reply: whether that call changed a row.
   *  `false` is not an error — two tabs open, or a second click. */
  changed?: boolean;
}

export const reviewApi = {
  get: async (): Promise<ReviewPayload> => {
    const { data } = await api.get<ReviewPayload>('/api/v1/review');
    return data;
  },

  /**
   * Confirm a guessed spending group.
   *
   * Household-scoped on the server: categories have no owner (D-20), so any
   * member may settle one. The account call below is the OPPOSITE rule, and a
   * 403 from it is expected rather than exceptional.
   */
  confirmCategory: async (categoryId: number): Promise<ReviewPayload> => {
    const { data } = await api.post<ReviewPayload>(
      `/api/v1/review/categories/${categoryId}/confirm`);
    return data;
  },

  /**
   * Confirm an inferred account type.
   *
   * *** A 403 HERE IS A NORMAL OUTCOME, NOT A BUG. *** The page shows every
   * household account on purpose — narrowing the read would put a row in the
   * list that its viewer cannot open (D-43) — but only the owner or an admin may
   * write one. The caller must render `error` from the response rather than
   * treating a rejection as a failed request.
   */
  confirmAccount: async (accountId: number): Promise<ReviewPayload> => {
    const { data } = await api.post<ReviewPayload>(
      `/api/v1/review/accounts/${accountId}/confirm`);
    return data;
  },
};

File: api/v1/review.py
"""Everything finPal guessed, in one place, so it can be confirmed or corrected.

*** THREE PATHS, AND THE GET IS THE ONLY ONE THE NAV NEEDS. *** `GET /review`
answers both "what is there to review?" and "is there anything at all?", so the
nav badge costs no extra round trip. Three endpoints answering three counts is
how a badge ends up flickering on every load.

*** A CONFIRM ANSWERS WITH THE WHOLE REFRESHED PAYLOAD, DELIBERATELY. *** The
obvious design returns `{'success': true}` and lets the client decrement its own
count — which is D-101 in miniature: the server owns the totals, and two clients
subtracting independently is two chances to disagree. Handing back the same shape
`GET` returns means the page re-renders from the server after every action and
cannot drift from it.

*** AND THE TWO REFUSALS ARE NOT THE SAME STATUS. ***

  * **403** — `ReviewNotPermitted`: the account belongs to another member. The
    page CORRECTLY shows it (reads stay household-wide, or D-43 comes back), so
    the client needs to say *why* clicking did nothing.
  * **200 with `changed: false`** — there was nothing to confirm. Two tabs open,
    or a second click. That is not an error and must not be dressed as one.

Collapsing those two is the defect this file was written to avoid: silence on a
row the page had just offered. `describe_refusal` (D-161) makes the same
distinction one layer down — two refusals that ask for different actions get two
answers.

Every handler returns `{'error': ...}` explicitly rather than letting restx shape
it: web-ui reads `err.response?.data?.error`, and restx answers `{'message': ...}`.
"""
import logging

from flask_jwt_extended import get_jwt_identity, jwt_required
from flask_restx import Namespace, Resource

from src.services.review.service import (ReviewNotPermitted, build_review,
                                         confirm_account, confirm_category)

logger = logging.getLogger(__name__)

ns = Namespace('review', description='Confirm or correct what finPal guessed')


def _payload(caller_id, changed=None):
    """The page, optionally with the outcome of the action that just ran."""
    body = build_review(caller_id)
    if changed is not None:
        body['changed'] = changed
    return body


@ns.route('/')
class Review(Resource):
    @jwt_required()
    def get(self):
        """Everything finPal guessed and the caller may answer for."""
        try:
            return _payload(get_jwt_identity()), 200
        except Exception:
            logger.exception('Failed to build the review page')
            return {'error': 'Could not load your review list'}, 500


class _Confirm(Resource):
    """The shared body of both confirmations.

    *** THE TWO CONFIRMATIONS USE OPPOSITE PERMISSION RULES AND THAT LIVES IN THE
    SERVICE, NOT HERE. *** A category is household property with no owner (D-20);
    an account is assignable to a member and is owner-or-admin. Only the HTTP
    translation is shared — putting the predicates here would give a reader two
    routes that look identical and hide the one difference that matters.
    """

    _confirm = None
    _what = 'row'

    @jwt_required()
    def post(self, item_id):
        caller_id = get_jwt_identity()
        try:
            changed = type(self)._confirm(caller_id, item_id)
        except ReviewNotPermitted as refusal:
            # 403, not 200: the page shows this row on purpose, so "nothing
            # happened" without a reason is the thing to avoid.
            return {'error': str(refusal)}, 403
        except Exception:
            logger.exception('Failed to confirm %s %s', self._what, item_id)
            return {'error': f'Could not confirm that {self._what}'}, 500
        return _payload(caller_id, changed=changed), 200


@ns.route('/categories/<int:item_id>/confirm')
class ConfirmCategory(_Confirm):
    _confirm = staticmethod(confirm_category)
    _what = 'category'


@ns.route('/accounts/<int:item_id>/confirm')
class ConfirmAccount(_Confirm):
    _confirm = staticmethod(confirm_account)
    _what = 'account'
